### PR TITLE
Enhanced offline

### PR DESCRIPTION
**Pull request by jakearchibald**
Adding things like an "update available" notification.

### CODE REVIEW

File: src/changelog.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+  "0.0.1": ["Pre-release"]
+}
```
Comment (jakearchibald) [inline src/changelog.json]: The first entry in the changelog is the current version. I want to find a way to display the changes between versions, but UI takes me a long time and I don't have enough of it before CDS. Will look into it after.

File: src/components/App/index.tsx [hunk 2]
```diff
@@ -12,6 +12,15 @@ import '../custom-els/LoadingSpinner';
...
 import Compress from '../compress';
 
+const compressPromise = import(
+  /* webpackChunkName: "main-app" */
+  '../compress',
+);
+const offlinerPromise = import(
+  /* webpackChunkName: "offliner" */
+  '../../lib/offliner',
+);
```
Comment (jakearchibald) [inline src/components/App/index.tsx]: I moved the service worker stuff into a lazy-loaded module to keep our first interaction size down.

File: src/lib/SnackBar/index.ts [hunk 3]
```diff
@@ -8,12 +8,9 @@ export interface SnackOptions {
 function createSnack(message: string, options: SnackOptions): [Element, Promise<string>] {
   const {
     timeout = 0,
-    actions = [],
+    actions = ['dismiss'],
```
Comment (jakearchibald) [inline src/lib/SnackBar/index.ts]: Snuck this in. Realised it doesn't make sense to have a snack that doesn't have a dismiss button.

File: webpack.config.js [hunk 4]
```diff
@@ -141,12 +142,6 @@ module.exports = function (_, env) {
           exclude: nodeModules,
           loader: 'ts-loader'
         },
-        {
-          test: /\.jsx?$/,
-          loader: 'babel-loader',
-          // Don't respect any Babel RC files found on the filesystem:
-          options: Object.assign(readJson('.babelrc'), { babelrc: false })
-        },
```
Comment (jakearchibald) [inline webpack.config.js]: Pretty sure we don't need this.

File: webpack.config.js [hunk 5]
```diff
@@ -242,30 +237,22 @@ module.exports = function (_, env) {
...
-      // Here we replace the constructor and message with a static throw, leaving the message to be DCE'd.
-      // This is useful since it shows the message in SourceMapped code when debugging.
-      isProd && new ReplacePlugin({
-        include: /babel-helper$/,
-        patterns: [{
-          regex: /throw\s+(?:new\s+)?((?:Type|Reference)?Error)\s*\(/g,
-          value: (s, type) => `throw 'babel error'; (`
-        }]
-      }),
-
```
Comment (jakearchibald) [inline webpack.config.js]: Pretty sure we don't need this.

File: src/lib/offliner.ts [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+  await updateReady(reg);
+
+  // Ask the user if they want to update.
+  const result = await showSnack('Update available', {
+    actions: ['reload', 'dismiss'],
+  });
+
+  // Tell the waiting worker to activate, this will change the controller and cause a reload (see
+  // 'controllerchange')
+  if (result === 'reload') skipWaiting();
```
Comment (surma) [inline src/lib/offliner.ts]: So just wondering: If I have 1 old and 1 new tab with Squoosh, and the new one shows me the reload prompt and I click reload, will the old tab reload as well or could we be breaking the old tab?

File: webpack.config.js [hunk 7]
```diff
@@ -19,6 +18,8 @@ function readJson (filename) {
   return JSON.parse(fs.readFileSync(filename));
 }
 
+const VERSION = Object.keys(readJson('./src/changelog.json'))[0];
```
Comment (surma) [inline webpack.config.js]: Should we rather read `package.json` for `VERSION`?

File: src/lib/offliner.ts (hunk 6 above)
Comment (jakearchibald) [reply src/lib/offliner.ts]: Hmm, it will reload all tabs.

File: webpack.config.js (hunk 7 above)
Comment (jakearchibald) [reply webpack.config.js]: Ohh good question. Do we want to force ourselves to write a changelog like this? I'll change it to use package json for this PR.

**Pull request merged**